File: HW1/10510_0317001.py
#This implementation doesn't detect invalid input
import sys

def is_higher_prec(op, stack_top):
    if (op == '+') and (
        stack_top == '*' or stack_top == '+'):
        return True
    elif op == '*' and stack_top == '*':
        return True
    else:
        return False

for line in sys.stdin:
    oprs = []
    post_exp = []
    for token in line:
        if token.isdigit():
            post_exp.append(token)
        else:
            if token == " " or token == "\n":
                pass
            elif token == ')':
                while oprs[-1] != '(':
                    post_exp.append(oprs[-1])
                    del oprs[-1]
                del oprs[-1]
            else:
                while len(oprs) and oprs[-1] != '(' and is_higher_prec(token, oprs[-1]):
                    # len(oprs) to prevent oprs[-1] out of range
                    post_exp.append(oprs[-1])
                    del oprs[-1]
                oprs.append(token)
    else:
        post_exp.extend(oprs[::-1])
        #"pop"(i.e. oprs[::-1]) ramaing operators in the stack
    L = []
    result = 0
    for token in post_exp:
        if token.isdigit():
            L.append(int(token))
        else:
            if token == '+':
                L[-2] = L[-2] + L[-1]
            else: # '*'
                L[-2] = L[-2] * L[-1]
            del L[-1]
                # operand<1> operand<2> operator --> operand, del last element
    print(L[0]%1000000007)
    
    

File: HW1/10510_0317001.cpp
#include <cstdio>
#include <cstring>
#include <cstdlib>
#include <cctype>

#define BUF_SIZE 10000

bool is_higher_prec(char op, char stack_top);

int main()
{
    char input[BUF_SIZE];
    char oprs[BUF_SIZE];
    char post_exp[BUF_SIZE];
    while(fgets(input, BUF_SIZE, stdin))
    {
        //printf("input: %s\n", input);
        int oprs_len = 0;
        int post_exp_len = 0;
        int len = strlen(input);
        for(int i = 0; i < len; i++)
        {
            //printf("------\n%c\n------\n", input[i]);
            if(isdigit(input[i]))
            {
                //printf("is digit! : %c\n", input[i]);
                post_exp[post_exp_len] = input[i];
                post_exp_len+=1;
            }
            else if(input[i] == ' ' || input[i] == '\n')
            {
                //printf("is space!\n");
                continue;
            }
            else if(input[i] == ')')
            {
                while( oprs[oprs_len - 1] != '(')
                {
                    post_exp[post_exp_len] = oprs[oprs_len - 1];
                    post_exp_len += 1;
                    oprs_len -= 1;
                }
                oprs_len -= 1;   
            }
            else
            {
                while( oprs_len && oprs[oprs_len - 1] != '(' 
                    && is_higher_prec(input[i], oprs[oprs_len - 1]) )
                // oprs[oprs_len-1] is "top" of the stack
                {
                    //printf("op ?!?!:%c\n", oprs[oprs_len]);
                    post_exp[post_exp_len] = oprs[oprs_len - 1];
                    post_exp_len += 1;
                    oprs_len -= 1;
                }
                oprs[oprs_len++] = input[i];
                //printf("LEN:%d\nContent:%s\n", oprs_len, oprs);
            }
            //printf("i iter's post_exp:%s\n", post_exp);
        }
        while(oprs_len)
        {
            post_exp[post_exp_len] = oprs[oprs_len - 1];
            post_exp_len += 1;
            oprs_len -= 1;
        }
        post_exp[post_exp_len] = '\0';
        //printf("what the fuck?? len: %d  post_exp:%s\n", post_exp_len, post_exp);




        long long result[10000], result_ptr = 0;
        //FUCK YOU long long
        for( auto c : post_exp)
        {
            if(isdigit(c))
            {
                result[result_ptr] = c - 48; // ASCII
                result_ptr += 1;
            }
            else
            {
                c == '+' ?
                    result[result_ptr - 2] = (result[result_ptr - 2] + result[result_ptr - 1])%1000000007
                    :result[result_ptr - 2] = (result[result_ptr - 2] * result[result_ptr - 1])%1000000007;
                result[result_ptr - 2] %= 1000000007;
                result_ptr -= 1;

            }
        }
        printf("%d\n", result[0]);

    }
}

bool is_higher_prec(char op, char stack_top)
{
    if (op == '+' && (stack_top == '*' or stack_top == '+'))
        return true;
    else if (op == '*' && stack_top == '*')
        return true;
    else
        return false;
}